### ISSUE TITLE
Support for mDNS?

### ISSUE THREAD

**Issue by infidel** [mirage/ocaml-dns#28]
I am interested in implementing a basic Multicast DNS resolver/responder (RFC 6762), mainly as a learning exercise (and only in my spare time). I had not heard of mDNS until reading the Cubieboard2 tutorials on openmirage.org, but I've been reading the RFCs since then. The top-level algorithms would be best implemented in a separate "mdns" package rather than in "dns", but most of the packets and other data structures will be shared with "dns", with minor differences.

Are you open to receiving patches to allow "dns" to support mDNS and perhaps DNS-SD?

As an example, DNS has a 16-bit resource record class field, while mDNS uses a 15-bit RR class plus a 1-bit "cache flush" bit (with the exception of some types like OPT). To start with, this requires a change to Dns.Packet.parse_rr to mask out bit 15 before calling int_to_rr_class, and to return the cache flush bit in the result.

**Comment by infidel** [comment]
OK taking your hint, I managed to get Jitsu going with normal DNS. It can resume/suspend a unikernel (I don't know how to get "shutdown" to work with a unikernel though). I understand that Jitsu is acting as a forwarding DNS resolver. Is it something that you would expect to use on the internet directly or is it more for use on a private LAN/WAN?

As for mDNS, which is for LANs, the concept of a "forwarder" doesn't exist but Jitsu could listen for queries of A records much like DNS. In the case where the unikernel started by Jitsu also implements a normal mDNS responder then it may be better if Jitsu did not even reply to the request, and just echoed the request once it expects that the unikernel is ready. This way Jitsu wouldn't even need to know the IP address of the unikernel, and maybe it could even use DHCP.

If the unikernel doesn't have its own mDNS responder then Jitsu could respond, and it could also perform the "probe" to ensure uniqueness of the hostname (if applicable).

Having Jitsu respond to DNS-SD PTR queries is another possibility.

**Comment by mor1** [comment]
On 7 January 2015 at 17:28, Luke Dunstan <notifications@github.com> wrote:

> OK taking your hint, I managed to get Jitsu going with normal DNS. It can
> resume/suspend a unikernel (I don't know how to get "shutdown" to work with
> a unikernel though). I understand that Jitsu is acting as a forwarding DNS
> resolver. Is it something that you would expect to use on the internet
> directly or is it more for use on a private LAN/WAN?
>
not sure we've really worked through all possible configurations. but at
least one is jitsu-as-unikernel-inetd, ie., a running jitsu would receive
dns query from the internet, and start/forward to unikernels as appropriate.

(or do i misunderstand the question? :)

> As for mDNS, which is for LANs, the concept of a "forwarder" doesn't exist
> but Jitsu could listen for queries of A records much like DNS. In the case
> where the unikernel started by Jitsu also implements a normal mDNS
> responder then it may be better if Jitsu did not even reply to the request,
> and just echoed the request once it expects that the unikernel is ready.
> This way Jitsu wouldn't even need to know the IP address of the unikernel,
> and maybe it could even use DHCP.
>
true; another option might be that the jitsu unikernel also implements a
DHCP server, so it can share state between the DNS and DHCP "threads".

-- 
Richard Mortier
mort@cantab.net

**Comment by MagnusS** [comment]
Mirage doesn't support shutdown at the moment I think (at least not on ARM), but you can use "destroy" to stop the unikernels with Jitsu ("-m destroy"). 

Jitsu tries to boot the unikernel as fast as possible to avoid missing the first TCP SYN from the client, so ideally it shouldn't have to wait for DHCP or mDNS. One approach could be to let Jitsu manage a pool of IP addresses (possibly obtained with DHCP) that it assigns to unikernels when they boot. If Jitsu knows the IP in advance it allows us to do optimizations such as [synjitsu](https://github.com/samoht/synjitsu), that caches received TCP SYNs for a specific IP while the unikernel is booting.

When Jitsu runs on a LAN and knows the domain name and IP of each unikernel, it would be great to be able to let Jitsu run mDNS on the unikernel's behalf and then only boot the unikernel when it is needed (as it does for normal DNS now). But we should probably continue the Jitsu discussion [here](https://github.com/MagnusS/jitsu) :-)

**Comment by infidel** [comment]
I've reviewed the code so far against a subset of RFC 6762, and rather than filing tons of issues here to keep track of what isn't implemented, I'm experimenting with annotating the RFC ("MUSTs" only so far):

http://infidel.github.io/ocaml-mdns/rfc6762_notes.html

(try to overlook the formatting)
The source is at: https://github.com/infidel/ocaml-mdns/blob/master/doc/rfc6762_notes.xml

Linking from the spec to the code is a temporary solution, because obviously when the code changes the links will break. It would be much better to do the reverse by putting code comments with references to the requirements, and extracting them automatically. Would that be acceptable though?

**Comment by dsheets** [comment]
That annotation is really cool! I think adopting a code -> rfc annotation system for this library is a good idea.

**Comment by avsm** [comment]
This is *really* nice!  

>  It would be much better to do the reverse by putting code comments with references to the requirements, and extracting them automatically. Would that be acceptable though?

Putting them in with the OCaml 4.02 support for annotations via ppx would work very well.  It would make mDNS require OCaml 4.02+, but this should be acceptable by now.

**Comment by infidel** [comment]
OK, yes I agree that attributes are the way to go. The very first attribute I tried to add triggered ocaml/camlp4#55 so that wasted a bit of time, but simple attributes do work. I understand the basic idea of extracting attributes from the AST (https://github.com/infidel/reqtrace/blob/master/src/ppx_reqtrace.ml) but I am not sure yet about the best way to invoke it for this use case.

The way that ppx extensions are intended to be invoked, by adding a -ppx option to the normal ocamlc step, has many advantages but is not a great fit because the goal is really to extract the attributes and write them to some separate file, whereas ppx extensions are meant be about transforming OCaml code before compiling it. It would work, but the data extraction would have to be a side-effect of the compilation, and it seems to be a pain to pass extra arguments to the ppx command, e.g. to specify the output path of the extraction module.

The alternative is to write a normal program that reads the OCaml source code independently of the compiler, but still using the same parsing library, which I think is how Merlin works. Advantages include:

- Simplicity: Easy to pass output file paths and other options on the command line. Don't have to emit a new AST.
- Efficiency: Could read all of the source files in one execution, instead of one file per run. Simpler to have one output file than multiple.
- Compilation doesn't depend on the reference extraction tool being installed.

Disadvantages:
- Redundancy: Some of the compiler options also need to be passed to the tool too (like .merlin), especially the camlp4 commands.
- OASIS doesn't know about it.
- More work to emit warnings and errors about attributes, compared to a ppx extension.
- More code overall.

Do you think it's more logical to make it independent of ocamlc, rather than an AST -> AST transformation plus side effects?

**Comment by infidel** [comment]
A third option seems rather attractive: post-process the *.cmt files using Cmt_format and TypedtreeIter. Attributes themselves are not signalled by the iterator but it's not difficult for e.g. enter_expression to extract one kind of attribute.

**Comment by dsheets** [comment]
`cmt` processing seems like the most robust, composable, and unintrusive approach.

**Comment by avsm** [comment]
`cmt` processing does seem robust, but it makes me wonder about just eliminating `camlp4` first. We're very close to be being able to do this; I believe just `type_conv` and `cstruct` need to change, and there are patches for the latter.  Otoh, that work will happen on its own timescale, and `cmt` processing would get this effort moving right now.  `cmt` processing is also how the codoc documentation works, so I wonder if there's some link there in the future with including more literate/spec documentation inline with interfaces.

**Comment by infidel** [comment]
OK, I have some code working with `cmt` files (borrowing a few bits from `codoc`). I could submit a PR to add some of the first references to the `ocaml-dns` code, but first I just want to know if my suggested attribute formats are acceptable:

1. I would like to distinguish between implementation code linked to requirements, and test code, so I was thinking of using an attribute at the top of each `.ml` file, like:

`[@@@reftype Impl]`
or
`[@@@reftype Test]`

The default could potentially be detected heuristically, e.g. based on the existence of the word "test" in the directory or file name. A default is probably not really necessary though.

2. For the simplest cases all the references in a particular `.ml` file would be to the same RFC, so this could be set at the top of the file too:

`[@@@reqdoc "RFC6762"]`

And then the rest of the source file can use concise references like `[@req "s6.7_p1_c1"]`.

3. Many source files will have to refer to multiple RFCs, which could be done individually:

`[@req "RFC6762:s6.7_p1_c1"]` (or something like that)

4. I find it hard to remember the RFC numbers. My idea to combine 2 and 3 above is:

`[@@@reqdoc let mdns = "RFC6762"]`
...
`[@req mdns "s6.7_p1_c1"]`

Some people consider it nice to have multiple ways to achieve the same thing, while others dislike it because it takes longer to learn. I'm OK with making (4) the only option if "one way" is the MirageOS philosophy.

**Comment by dsheets** [comment]
```
[@@@reqdoc let mdns = "RFC6762"]
```

and

```
[@req mdns "s6.7_p1_c1"]
```

seem equivalent to

```
[@req "RFC6762" "s6.7_p1_c1"]
```

I like the rest of the design in general except:

1. What does `req` stand for? "Request (for Comments)"? What if we use this for other standards families?

2. Related to 1., should only the number/name of the RFC be given or could a URI be supplied so other documents can be referenced?

So maybe something like:

```
[@@@specdoc let mdns = rfc 6762]
```

thus

```
[@ref (rfc 6762) "s6.7_p1_c1"]
```

but then also

```
[@@@specdoc let my_spec = uri "http://example.net/coolest_standard_ever"]
```

**Comment by infidel** [comment]
Thanks. I agree, [@req "RFC6762" "s6.7_p1_c1"] is more logical as an equivalent.

"req" is short for "requirement", not meant to be specific to RFCs (hence reqtrace short for requirements traceability).

While the identification of requirements specification documents was mainly intended to be interpreted by the extraction tool, I think the idea of referring to non-RFC specifications with a URI is good. Using (rfc 6762) instead of a string literals is better too.

**Comment by infidel** [comment]
Note that my PR doesn't use `[@ref (rfc 6762) "..."]` syntax, nor floating attributes `[@@@specdoc mdns (rfc 6762)]` because neither of those work with camlp4.

This is also noted at: https://github.com/infidel/reqtrace

**Comment by yomimono** [comment]
I believe the mdns work is merged, so I'm closing this issue.  Please feel free to re-open it and update us on the status if I'm incorrect!
